libavcodec/libaribcaption.c

```diff
@@ -0,0 +1,1141 @@
+        return AVERROR_EXTERNAL;
+    }
+    aribcc_decoder_set_replace_msz_fullwidth_ascii(ctx->decoder,
+                                                   ctx->replace_fullwidth_ascii);
+
+    switch (ctx->subtitle_type) {
+    case SUBTITLE_ASS:
+        ret = set_ass_header(ctx);
+        if (ret != 0) {
+            av_log(avctx, AV_LOG_ERROR, "Failed to set ASS header: %s\n",
+                                        av_err2str(ret));
+            aribcaption_close(avctx);
+            return ret;
+        }
+        break;
+
+    case SUBTITLE_BITMAP:
+        if(!(ctx->renderer = aribcc_renderer_alloc(ctx->context))) {
+            av_log(avctx, AV_LOG_ERROR, "Failed to alloc libaribcaption renderer.\n");
+            aribcaption_close(avctx);
+            return AVERROR_EXTERNAL;
+        }
+        if(!aribcc_renderer_initialize(ctx->renderer,
+                                       ARIBCC_CAPTIONTYPE_CAPTION,
+                                       ARIBCC_FONTPROVIDER_TYPE_AUTO,
+                                       ARIBCC_TEXTRENDERER_TYPE_AUTO)) {
+            av_log(avctx, AV_LOG_ERROR, "Failed to initialize libaribcaption renderer.\n");
+            aribcaption_close(avctx);
+            return AVERROR_EXTERNAL;
+        }
+        estimate_video_frame_size(ctx);
+        if (!aribcc_renderer_set_frame_size(ctx->renderer,
+                                            ctx->frame_width, ctx->frame_height)) {
+            av_log(ctx, AV_LOG_ERROR,
+                   "aribcc_renderer_set_frame_size() returned with error.\n");
+            aribcaption_close(avctx);
+            return AVERROR_EXTERNAL;
+        }
+        if (!(ctx->clut = av_mallocz(AVPALETTE_SIZE))) {
+            aribcaption_close(avctx);
+            return AVERROR(ENOMEM);
+        }
+
+        aribcc_renderer_set_storage_policy(ctx->renderer, ARIBCC_CAPTION_STORAGE_POLICY_MINIMUM, 0);
+        aribcc_renderer_set_replace_drcs(ctx->renderer, ctx->replace_drcs);
+        aribcc_renderer_set_force_stroke_text(ctx->renderer, ctx->force_stroke_text);
+        aribcc_renderer_set_force_no_background(ctx->renderer, ctx->ignore_background);
+        aribcc_renderer_set_force_no_ruby(ctx->renderer, ctx->ignore_ruby);
+        aribcc_renderer_set_stroke_width(ctx->renderer, ctx->stroke_width);
+        if (ctx->font) {
+            int is_nomem = 0;
+            size_t count = 0;
+            const char **font_families = NULL;
+            const char *fonts = ctx->font;
+
+            while (!is_nomem && *fonts) {
+                const char **ff = av_realloc_array(font_families, count + 1, sizeof(*font_families));
+                if (!ff)
+                    is_nomem = 1;
+                else {
+                    font_families = ff;
+                    ff[count++] = av_get_token(&fonts, ",");
+                    if (!ff[count - 1])
+                        is_nomem = 1;
+                    else if (*fonts)
+                        fonts++;
+                }
```

Comment:
If you get an allocation failure, you should return AVERROR(ENOMEM), rather than attempting to continue in a degraded state.

Reply:
This does't continue in a degraded state, but return `AVERROR(ENOMEM)` after clean up.

```c
            while (count)
                av_freep(&font_families[--count]);
            av_freep(&font_families);
            if (is_nomem) {
                aribcaption_close(avctx);
                return AVERROR(ENOMEM);
            }
```

If such clean up is not needed (i.e., automatically freed), I can simply return with `AVERROR(ENOMEM)`.

Reply:
You can add a label (e.g. `fail:`) and `goto` it on failure.

Reply:
It is a matter of preference if it is not specified in coding rule.
I prefer to keep current way because same clean up routine runs regardless of error or not.

Reply:
Instead of `while (!is_nomem && *fonts)` loop, I can change to add `break;` at allocation failure to quit from loop.

Reply:
Sure, that's a bit cleaner.